code/datums/status_effects/xeno_buffs.dm

```diff
@@ -24,6 +27,335 @@
+	var/datum/beam/current_beam
+	/// Used to prevent duplicate heals in the case of two damage sources.
+	var/heal_count
+
+/datum/status_effect/stacking/essence_link/on_creation(mob/living/new_owner, stacks_to_apply, mob/living/carbon/link_target)
+	link_owner = new_owner
+	src.link_target = link_target
+	essence_link_action = link_owner.actions_by_path[/datum/action/xeno_action/activable/essence_link]
+	ADD_TRAIT(link_owner, TRAIT_ESSENCE_LINKED, TRAIT_STATUS_EFFECT(id))
+	ADD_TRAIT(link_target, TRAIT_ESSENCE_LINKED, TRAIT_STATUS_EFFECT(id))
+	RegisterSignal(link_owner, COMSIG_MOB_DEATH, .proc/handle_death)
+	RegisterSignal(link_target, COMSIG_MOB_DEATH, .proc/handle_death)
+	RegisterSignal(link_owner, COMSIG_MOVABLE_MOVED, .proc/handle_dist)
+	RegisterSignal(link_target, COMSIG_MOVABLE_MOVED, .proc/handle_dist)
+	RegisterSignal(link_owner, COMSIG_XENOMORPH_RESIN_JELLY_APPLIED, .proc/share_jelly)
+	RegisterSignal(link_target, COMSIG_XENOMORPH_RESIN_JELLY_APPLIED, .proc/share_jelly)
+	RegisterSignal(link_owner, COMSIG_XENOMORPH_BRUTE_DAMAGE, .proc/share_heal)
+	RegisterSignal(link_target, COMSIG_XENOMORPH_BRUTE_DAMAGE, .proc/share_heal)
+	RegisterSignal(link_owner, COMSIG_XENOMORPH_BURN_DAMAGE, .proc/share_heal)
+	RegisterSignal(link_target, COMSIG_XENOMORPH_BURN_DAMAGE, .proc/share_heal)
+	toggle_link(TRUE)
+	to_chat(link_target, "[link_owner] has established an Essence Link with you. Stay within [DRONE_ESSENCE_LINK_RANGE] tiles to maintain it.")
+	return ..()
+
+/datum/status_effect/stacking/essence_link/add_stacks(stacks_added)
+	. = ..()
+	essence_link_action.update_button_icon()
+	link_owner.balloon_alert(link_owner, "Attunement: [stacks]/[max_stacks]")
+	COOLDOWN_START(src, attunement_increase, attunement_cooldown)
+	update_beam()
+
+/datum/status_effect/stacking/essence_link/on_remove()
+	. = ..()
+	link_owner.balloon_alert(link_owner, "Essence Link cancelled")
+	link_target.balloon_alert(link_target, "Essence Link cancelled")
+	toggle_link(FALSE)
+	essence_link_action.end_ability()
+	REMOVE_TRAIT(link_owner, TRAIT_ESSENCE_LINKED, TRAIT_STATUS_EFFECT(id))
+	REMOVE_TRAIT(link_target, TRAIT_ESSENCE_LINKED, TRAIT_STATUS_EFFECT(id))
+	UnregisterSignal(link_owner, list(COMSIG_MOB_DEATH, COMSIG_MOVABLE_MOVED, COMSIG_XENOMORPH_RESIN_JELLY_APPLIED, COMSIG_XENOMORPH_BRUTE_DAMAGE, COMSIG_XENOMORPH_BURN_DAMAGE))
+	UnregisterSignal(link_target, list(COMSIG_MOB_DEATH, COMSIG_MOVABLE_MOVED, COMSIG_XENOMORPH_RESIN_JELLY_APPLIED, COMSIG_XENOMORPH_BRUTE_DAMAGE, COMSIG_XENOMORPH_BURN_DAMAGE))
+
+/datum/status_effect/stacking/essence_link/tick()
+	var/remaining_health = link_target.maxHealth - (link_target.getBruteLoss() + link_target.getFireLoss())
+	var/heal_amount = round(link_target.maxHealth * (stacks * DRONE_ESSENCE_LINK_REGEN))
+	var/heal_remainder = round(max(0, heal_amount - link_target.getBruteLoss()))
+	var/plasma_cost = round(heal_amount * 2.5)
+	if(stacks < max_stacks && COOLDOWN_CHECK(src, attunement_increase))
```

Comment:
These vars should either not exist (used once), or declared later (right before using them)

Reply:
Donezo.